gensim/models/indexer.py

```diff
@@ -0,0 +1,246 @@
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        if type(model) is Doc2Vec:
+            self._init_doc2vec_(model)
+        elif type(model) is Word2Vec:
+            self._init_word2vec_(model)
+        else:
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        self.start_indexing()
+
+    def _init_doc2vec_(self, model):
+        docvecs = model.docvecs
+        docvecs.init_sims()
+        size = len(docvecs.doctag_syn0norm)
+
+        for i in range(size):
+            self.add_item(docvecs.offset2doctag[i], docvecs.doctag_syn0norm[i])
+
+    def _init_word2vec_(self, model):
+        model.init_sims()
+        size = len(model.syn0norm)
+
+        for i in range(size):
+            self.add_item(model.index2word[i], model.syn0norm[i])
+
+    def get_item(self, label):
+        """
+        Get an item's vector by its label
+
+        Args:
+            label: The label
+
+        Returns:
+            The item's vector if have. Otherwise, returns None
+        """
+        pass
+
+    def add_item(self, label, vector):
+        """
+        Add an item to index
+
+        Args:
+            label: the label of the item (must be unique).
+            vector: the item's vector.
+        """
+        pass
+
+    def start_indexing(self):
+        """
+        Start the indexing operation.
+        This method is supposed to run after all items are added
+        It may take time to finish.
+        """
+        pass
+
+    def get_nearest_items(self, vector, top_n=10):
+        """
+        Get nearest items for an item described by its vector
+
+        Args:
+            vector: The vector
+            top_n: Number of nearest items to get
+
+        Returns:
+            A list of tuples (label, similarity) for the nearest ones
+        """
+        pass
+
+    def save(self, file_name):
+        """
+        Dump internal data into disk
+
+        Args:
+            file_name:   Output file path
+
+        Returns:
+            Returns true on success. Otherwise, returns false
+        """
+        pass
+
+    def load(self, file_name):
+        """
+        Load previously dumped info
+
+        Args:
+            file_name:  Input file paht
+
+        Returns:
+            Returns True on success. Otherwise, returns False
+        """
+        pass
+
+
+class AnnoyIndexer(Indexer):
+    """
+    An Annoy-based indexer.
+    For more information about Annoy, please visit https://github.com/spotify/annoy
+    """
+
+    def __init__(self, model=None, **kwargs):
+        """
+        Constructor
+
+        Args:
+            features_num: Number of features
+            tree_size: Number of trees. More trees gives higher precision when querying.
+
+        Raises:
+            TypeError: not provide enough parameters 
+        """
+        if "features_num" not in kwargs:
```

Comment:
Why not simply name the parameter inside `__init__(model=None, feature_num=None)` ? Looks identical but more idiomatic.